[RemoteMirror] Fix handling of zero-sized payloads and extra inhabitant calculations

tbkka opened the pull request:
This PR aligns RemoteMirror more closely with the compiler's handling of enums that contain zero-sized payloads, which should resolve a number of issues with RemoteMirror incorrectly reflecting enum cases and/or measuring the layout of structures containing enums.

Background: Enum cases that have zero-sized payloads are handled differently from other payload-bearing cases:

1. For layout purposes, they're treated as non-payload cases.  This can cause a multi-payload enum to actually get represented in memory as a single-payload or non-payload enum.

2. However, zero-sized payloads are still considered for extra inhabitant calculations.  Since they have no extra inhabitants, this tends to cause such enums to also not expose extra inhabitants to containing enums.

This commit makes several change to how RemoteMirror determines enum layout:

* The various "*EnumTypeInfo" classes now represent layout mechanisms -- as described in (1) above, this can differ from the source code concept.

* An Enum "kind" is separately computed to reflect the source code concept; this ensures that the dumped type information reflects the source code.

* For single-payload and no-payload _layouts_, the extra inhabitant calculation has been adjusted to ensure that zero-sized payloads are correctly considered.

Resolves: rdar://92945673

tbkka committed:
This PR aligns RemoteMirror more closely with the compiler's handling of enums
that contain zero-sized payloads, which should resolve a number of issues with
RemoteMirror incorrectly reflecting enum cases and/or measuring the layout of
structures containing enums.

Background: Enum cases that have zero-sized payloads are handled
differently from other payload-bearing cases.

1. For layout purposes, they're treated as
   non-payload cases.  This can cause an MPE to
   actually get represented in memory as a single-payload
   or non-payload enum.

2. However, zero-sized payloads are still considered for
   extra inhabitant calculations.  Since they have no
   extra inhabitants, this tends to cause such enums to
   also not expose extra inhabitants to containing enums.

This commit makes several change to how RemoteMirror determines
enum layout:

* The various "*EnumTypeInfo" classes now represent
  layout mechanisms -- as described in (1) above,
  this can differ from the source code concept.

* An Enum "kind" is separately computed to reflect the
  source code concept; this ensures that the dumped
  type information reflects the source code.

* For single-payload and no-payload _layouts_,
  the extra inhabitant calculation has been adjusted
  to ensure that zero-sized payloads are correctly
  considered.

Resolves: rdar://92945673

tbkka commented:
TODO:  I still need to verify the tests on 32-bit platforms.

tbkka commented:
@swift-ci Please test

mikeash commented on include/swift/Reflection/TypeLowering.h:
```diff
@@ -239,6 +239,14 @@ class EnumTypeInfo : public TypeInfo {
     return std::count_if(Cases.begin(), Cases.end(),
                          [](const FieldInfo &Case){return Case.TR != 0;});
   }
+  unsigned getNumNonEmptyPayloadCases() const {
```
Looks like this is only used in asserts, does this need any magic to avoid an unused-function warning in a no-asserts build?

mikeash reviewed (approved):
Nice!

tbkka commented:
@swift-ci Please test

tbkka commented:
Argh.  One of the tests is crashing without explanation on CI but not locally.  This could take a while.

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test macOS Platform

tbkka committed:
Match the earlier behavior when we lack typeinfo
for a particular payload.  This should fix a test
crash that occurred on x86_64 with CF payloads.

tbkka commented:
@swift-ci Please test

tbkka commented:
Ummm.... GitHub somehow added 7 new reviewers?  No idea why?

I figured out why one of the tests was crashing, but still no idea why it was _only_ crashing on x86.  🤷

tbkka commented:
Linux failed in an LLDB Regex test (likely unrelated?), and macOS testing just failed to start.  Trying again...

tbkka commented:
@swift-ci Please test

tbkka commented:
Ah.  The Linux LLDB failure is _not_ unrelated.  More work to come.

tbkka committed:
Consistently treat payloads with missing/invalid typeinfo as non-empty.

tbkka commented:
@swift-ci Please test

compnerd commented on include/swift/RemoteInspection/TypeLowering.h:
```diff
@@ -239,6 +239,17 @@ class EnumTypeInfo : public TypeInfo {
+                           return Case.TR != 0
+                             && (Case.TI.getKind() == TypeInfoKind::Invalid
+                                 || Case.TI.getSize() > 0);
+                         });
```
I wonder if it makes sense to extract the various predicates used for classifying the payload information into standalone functions that are marked as `inline` and put into a shared header.

tbkka replied:
Good idea, but I'm not yet entirely sure what those predicates would be.  The enum support here needs a big round of refactoring.  Slava aspires to someday merge this with the type layout machinery being used in the compiler, but I fear that's much further off.
